src/dhtproto/client/legacy/internal/helper/mirror/model/MirrorBase.d

```diff
@@ -0,0 +1,257 @@
+/*******************************************************************************
+
+    DHT channel mirror helper base class. Defines the basic API.
+
+    API for a helper class that reads records from a DHT channel as they are
+    modified. A mirror reads using two techniques:
+        1. A DHT Listen request to receive records as they are modified.
+        2. A periodically activated DHT GetAll request to ensure that the Listen
+           request does not miss any records due to connection errors, etc.
+
+    The actual request assignment and handling are abstract.
+
+    Copyright:
+        Copyright (c) 2011-2018 sociomantic labs GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module dhtproto.client.legacy.internal.helper.mirror.model.MirrorBase;
+
+import ocean.transition;
+
+import dhtproto.client.DhtClient;
+
+/*******************************************************************************
+
+    Channel mirror API abstract base class.
+
+    Template params:
+        Dht = type of DHT client (must be derived from DhtClient and contain the
+            RequestScheduler plugin)
+
+*******************************************************************************/
+
+abstract public class MirrorBase ( Dht : DhtClient )
+{
+    /// This helper only works in conjunction with the request scheduler plugin,
+    /// so we assert that that exists in the client.
+    static assert(Dht.HasPlugin!(DhtClient.RequestScheduler),
+        "MirrorBase requires a DHT client with the RequestScheduler plugin");
+
+    /// DHT client used to access DHT.
+    protected Dht dht;
+
+    /// Name of DHT channel to mirror.
+    protected istring channel_;
```

Comment:
Is there any reason for this to be protected and not private?

Reply:
To maintain the API. User code may exist that accesses `ChannelMirror.channel_`(https://github.com/gavin-norman-sociomantic/dhtproto-tsunami/blob/330ebffb8d2101b7884ede16db9d66c3a48f5dbe/src/dhtproto/client/legacy/internal/helper/ChannelMirror.d#L284) in a derived class. `Mirror` is supposed to be a drop-in replacement for `ChannelMirror`, so the field must have the same protection attribute.

Reply:
OK, that's a valid reason.